crates/compute/src/cpu/layer.rs

```diff
@@ -424,6 +424,53 @@ impl<F: TowerTop> ComputeLayerExecutor<F> for CpuLayerExecutor<F> {
 
 		Ok(())
 	}
+
+	fn pairwise_product_reduce(
+		&mut self,
+		input: <Self::DevMem as ComputeMemory<F>>::FSlice<'_>,
+		round_outputs: &mut [<Self::DevMem as ComputeMemory<F>>::FSliceMut<'_>],
+	) -> Result<(), Error> {
+		if input.len().count_ones() != 1 {
```

Comment:
Maybe make sense to do something like this 
`let Some(log_num_inputs) = strict_log_2(input.len()) else {return Err(...)} `